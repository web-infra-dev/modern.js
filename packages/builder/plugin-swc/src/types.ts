import type {
  EnvConfig,
  Extensions,
  JsMinifyOptions,
  ReactConfig,
  TransformConfig,
} from '@modern-js/swc-plugins';
import type { lodash as _ } from '@modern-js/utils';

export type {
  Output,
  TransformConfig,
  JsMinifyOptions,
  TerserCompressOptions,
} from '@modern-js/swc-plugins';

export type OuterExtensions = Omit<
  Extensions,
  'ssrLoaderId' | 'configRoutes'
> & {
  /**
   * @deprecated
   */
  // for backwards compatibility
  modernjsSsrLoaderId?: boolean;
};

export interface ObjPluginSwcOptions<T extends 'inner' | 'outer' = 'inner'>
  extends TransformConfig {
  presetReact?: ReactConfig;
  presetEnv?: EnvConfig;

  jsMinify?: boolean | JsMinifyOptions;
  cssMinify?: boolean | CssMinifyOptions;

  extensions?: T extends 'inner' ? Extensions : OuterExtensions;

  overrides?: Override<T>[];
}

export type Override<T extends 'inner' | 'outer'> = Omit<
  ObjPluginSwcOptions<T>,
  'test' | 'include' | 'exclude'
> & {
  test?: RegExp;
  include?: RegExp[];
  exclude?: RegExp[];
};

export type FnPluginSwcOptions = (
  config: TransformConfig,
  utilities: Utilities,
) => void | TransformConfig;

export type PluginSwcOptions<T extends 'inner' | 'outer' = 'inner'> =
  | ObjPluginSwcOptions<T>
  | FnPluginSwcOptions;

interface Utilities {
  mergeConfig: typeof _.merge;
  setConfig: typeof _.set;
}

export interface CssMinifyOptions {
  sourceMap?: boolean;
  inlineSourceContent?: boolean;
}
