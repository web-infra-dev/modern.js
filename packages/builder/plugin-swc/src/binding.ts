import { Compiler as RawCompiler, Output } from '@modern-js/swc-plugins';
import { TransformConfig } from './types';

export {
  minify,
  minifySync,
  minifyCss,
  minifyCssSync,
} from '@modern-js/swc-plugins';

export class Compiler extends RawCompiler {
  config: TransformConfig;

  constructor(finalConfig: TransformConfig) {
    super(finalConfig);
    this.config = finalConfig;
  }
}

export function transformSync(
  config: Required<TransformConfig>,
  filename: string,
  code: string,
  map?: string,
): Output {
  const compiler = new Compiler(config);

  return compiler.transformSync(filename, code, map);
}

export function transform(
  config: Required<TransformConfig>,
  filename: string,
  code: string,
  map?: string,
): Promise<Output> {
  let compiler;
  try {
    compiler = new Compiler(config);
  } catch (e) {
    throw new Error(`[builder-plugin-swc] Failed to initialize config: \n${e}`);
  }
  return compiler.transform(filename, code, map);
}
