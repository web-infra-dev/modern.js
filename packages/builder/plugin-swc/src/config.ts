import { merge } from '@modern-js/utils/lodash';
import { TransformConfig, TransformConfigNapi } from '@modern-js/swc-plugins';

export type { TransformConfig } from '@modern-js/swc-plugins';

export type PartialRecursive<O> = O extends Record<any, any>
  ? O extends any[]
    ? O
    : { [k in keyof O]?: PartialRecursive<O[k]> }
  : O;

/// default swc configuration
function getDefaultSwcConfig(): Required<TransformConfig> {
  return {
    swc: {
      cwd: process.cwd(),
      jsc: {
        target: 'es5',
        externalHelpers: true,
        parser: {
          tsx: true,
          syntax: 'typescript',
        },
        transform: {
          react: {
            runtime: 'automatic',
          },
        },
      },
      minify: false, // for loader, we don't need to minify, we do minification using plugin
      module: {
        type: 'es6',
        ignoreDynamic: true, // let bundler handle dyn import
      },
      sourceMaps: true,
      env: {
        targets: '',
        mode: 'usage',
      },
      test: '',
      exclude: [],
      inlineSourcesContent: true,
    },
    extensions: {},
  };
}

export function normalizeConfig(
  opt: TransformConfig,
): Required<TransformConfig> {
  // complete default config
  const config: Required<TransformConfig> = merge(
    getDefaultSwcConfig(),
    opt,
  ) as unknown as Required<TransformConfig>;

  const { swc } = config;
  if (!swc.filename) {
    swc.filename = 'anonymous';
  }
  return config;
}

export function toBindingConfig(config: TransformConfig): TransformConfigNapi {
  return {
    swc: JSON.stringify(config.swc),
    extensions: config.extensions,
  };
}

export function defineConfig(options: TransformConfig): TransformConfig {
  return options;
}
