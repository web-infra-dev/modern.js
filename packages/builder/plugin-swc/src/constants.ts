import * as path from 'path';

export const CORE_JS_PATH = require.resolve('core-js/package.json');
export const SWC_HELPERS_PATH = require.resolve('@swc/helpers/package.json');

export const CORE_JS_DIR_PATH = path.dirname(CORE_JS_PATH);
export const SWC_HELPERS_DIR_PATH = path.dirname(SWC_HELPERS_PATH);

export const JS_REGEX = /\.js$/;
export const CSS_REGEX = /\.css$/;
