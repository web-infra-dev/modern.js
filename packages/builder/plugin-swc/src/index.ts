export { default as SwcLoader } from './loader';
export { builderPluginSwc, PluginSwc } from './plugin';
export * from './binding';
export * from './types';

/// NOTE:
/// Polyfill async function or generator, should install regenerator-runtime manually
