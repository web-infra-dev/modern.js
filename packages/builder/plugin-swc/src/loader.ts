import { getBrowserslist } from '@modern-js/utils';
import type { LoaderContext, LoaderDefinitionFunction } from 'webpack';
import { Options as SwcOptions } from '@modern-js/swc-plugins';
import { Compiler } from './binding';
import { TransformConfig } from './types';

function setReactDevMode(
  swc: SwcOptions,
  mode: 'development' | 'production' | 'none',
) {
  if (!swc.jsc) {
    swc.jsc = {};
  }
  if (!swc.jsc.transform) {
    swc.jsc.transform = {};
  }
  if (!swc.jsc.transform.react) {
    swc.jsc.transform.react = {};
  }

  const { react } = swc.jsc.transform;
  react.development = react.development ?? mode === 'development';
  react.refresh = react.refresh ?? mode === 'development';
}

function normalizeLoaderOption(
  options: TransformConfig,
  ctx: LoaderContext<TransformConfig>,
) {
  const enableSourceMap = ctx.sourceMap;

  if (enableSourceMap) {
    options.sourceMaps = true;
  }

  if (options.env && !options.env.targets) {
    options.env.targets = getBrowserslist(options.cwd || process.cwd());
  }

  if (
    !options.jsc?.transform?.react ||
    options.jsc.transform.react.development === undefined
  ) {
    setReactDevMode(options, ctx.mode);
  }

  // disable unnecessary config searching
  // all config should be explicitly set
  options.swcrc = false;
}

export function createLoader(): LoaderDefinitionFunction {
  const compilers = new Map<Record<string, any>, Compiler>();

  function getCompiler(options: Required<TransformConfig>) {
    if (compilers.has(options)) {
      return compilers.get(options)!;
    }

    const compiler = new Compiler(options);
    compilers.set(options, compiler);
    return compiler;
  }

  return function SwcLoader(code, map) {
    const resolve = this.async();
    const filename = this.resourcePath;

    const options = this.getOptions();
    normalizeLoaderOption(options, this);
    const compiler = getCompiler(options as Required<TransformConfig>);
    compiler
      .transform(
        filename,
        code,
        typeof map === 'object' ? JSON.stringify(map) : map,
      )
      .then(result => {
        resolve(null, result.code, result.map);
      })
      .catch(err => {
        resolve(err as Error);
      });
  };
}

export default createLoader();
