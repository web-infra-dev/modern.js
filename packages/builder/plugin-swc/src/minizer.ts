import { webpack } from '@modern-js/builder-webpack-provider';
import { merge } from '@modern-js/utils/lodash';
import { chalk, logger } from '@modern-js/utils';
import { CssMinifyOptions, JsMinifyOptions, Output } from './types';
import { minify, minifyCss } from './binding';
import { CSS_REGEX, JS_REGEX } from './constants';

const defaultMinifyOptions: JsMinifyOptions = {
  compress: {},
  mangle: true,
};

export interface NormalizedSwcMinifyOption {
  jsMinify?: JsMinifyOptions;
  cssMinify?: CssMinifyOptions;
}

const normalize = <T>(
  v: T | boolean | undefined,
  defaultValue: T,
): T | undefined => {
  if (v === true || v === undefined) {
    return defaultValue;
  } else if (v === false) {
    return undefined;
  } else {
    return v;
  }
};

export class SwcMinimizerPlugin {
  private readonly minifyOptions: NormalizedSwcMinifyOption;

  private name: string = 'swc-minimizer-plugin';

  constructor(
    options: {
      jsMinify?: boolean | JsMinifyOptions;
      cssMinify?: boolean | CssMinifyOptions;
    } = {},
  ) {
    this.minifyOptions = {
      jsMinify: merge(defaultMinifyOptions, normalize(options.jsMinify, {})),
      cssMinify: normalize(options.cssMinify, {}),
    };
  }

  apply(compiler: webpack.Compiler): void {
    const meta = JSON.stringify({
      name: 'swc-minify',
      options: this.minifyOptions,
    });

    compiler.hooks.compilation.tap(this.name, async compilation => {
      const { Compilation } = compiler.webpack;
      const { devtool } = compilation.options;
      const { jsMinify, cssMinify } = this.minifyOptions;

      const enableMinify =
        typeof devtool === 'string'
          ? devtool.includes('source-map')
          : Boolean(devtool);
      const inlineSourceContent =
        typeof devtool === 'string' && devtool.includes('inline');

      if (jsMinify) {
        jsMinify.sourceMap = enableMinify;
        jsMinify.inlineSourcesContent = inlineSourceContent;
      }

      if (cssMinify) {
        cssMinify.sourceMap = enableMinify;
        cssMinify.inlineSourceContent = inlineSourceContent;
      }

      compilation.hooks.chunkHash.tap(this.name, (_, hash) =>
        hash.update(meta),
      );

      compilation.hooks.processAssets.tapPromise(
        {
          name: this.name,
          stage: Compilation.PROCESS_ASSETS_STAGE_OPTIMIZE_SIZE,
        },
        async () => {
          try {
            await this.updateAssets(compilation);
          } catch (e) {
            compilation.errors.push(
              new compiler.webpack.WebpackError(`[SWC Minify]: ${e}`),
            );
          }
        },
      );
    });
  }

  async updateAssets(compilation: webpack.Compilation): Promise<void[]> {
    const cache = compilation.getCache(this.name);

    const { SourceMapSource, RawSource } = compilation.compiler.webpack.sources;
    const assets = compilation
      .getAssets()
      .filter(
        asset =>
          !asset.info.minimized &&
          (JS_REGEX.test(asset.name) || CSS_REGEX.test(asset.name)),
      );

    const assetsWithCache = await Promise.all(
      assets.map(async ({ name, info, source }) => {
        const eTag = cache.getLazyHashedEtag(source);
        const cacheItem = cache.getItemCache(name, eTag);
        return {
          name,
          info,
          source,
          cacheItem,
        };
      }),
    );

    const { cssMinify, jsMinify } = this.minifyOptions;
    return Promise.all(
      assetsWithCache.map(async asset => {
        const cache = await asset.cacheItem.getPromise<{
          minifiedSource: InstanceType<
            typeof SourceMapSource | typeof RawSource
          >;
        }>();

        let minifiedSource = cache ? cache.minifiedSource : null;

        if (!minifiedSource) {
          const { source, map } = asset.source.sourceAndMap();
          let minifyResult: Output | undefined;
          let needSourceMap = false;
          const filename = asset.name;

          if (jsMinify && filename.endsWith('.js')) {
            needSourceMap = jsMinify.sourceMap!;
            minifyResult = await minifyWithTimeout(filename, () => {
              return minify(filename, source.toString(), jsMinify);
            });
          } else if (cssMinify && filename.endsWith('.css')) {
            needSourceMap = cssMinify.sourceMap!;
            minifyResult = await minifyWithTimeout(filename, () => {
              return minifyCss(filename, source.toString(), cssMinify);
            });
          }

          if (minifyResult) {
            minifiedSource =
              needSourceMap && minifyResult.map
                ? new SourceMapSource(
                    minifyResult.code,
                    asset.name,
                    minifyResult.map,
                    source.toString(),
                    map,
                    true,
                  )
                : new RawSource(minifyResult.code || '');
          }
        }

        if (minifiedSource) {
          await asset.cacheItem.storePromise({
            minifiedSource,
          });

          compilation.updateAsset(asset.name, minifiedSource, {
            ...asset.info,
            minimized: true,
          });
        }
      }),
    );
  }
}

/**
 * Currently SWC minify is not stable as we expected, there is a
 * change that it can never ends, so add a warning if it hangs too long.
 */
function minifyWithTimeout(
  filename: string,
  minify: () => Promise<Output>,
): Promise<Output> {
  const timer = setTimeout(() => {
    logger.warn(
      `SWC minimize has running for over 180 seconds for a single file: ${filename}\n
It is likely that you've encountered a ${chalk.red(
        'SWC internal bug',
      )}, please contact us at https://github.com/web-infra-dev/modern.js/issues`,
    );
  }, 180_000);

  const outputPromise = minify();

  outputPromise.finally(() => {
    clearTimeout(timer);
  });

  return outputPromise;
}
