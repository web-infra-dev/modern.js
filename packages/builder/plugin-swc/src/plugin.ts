import * as path from 'path';
import type {
  BuilderPluginAPI,
  BuilderPlugin,
} from '@modern-js/builder-webpack-provider';
import { JS_REGEX, TS_REGEX, mergeRegex } from '@modern-js/builder-shared';
import type { PluginSwcOptions, TransformConfig } from './types';
import {
  applyPluginConfig,
  checkUseMinify,
  isDebugMode,
  removeUselessOptions,
} from './utils';
import { SwcMinimizerPlugin } from './minizer';

const PLUGIN_NAME = 'builder-plugin-swc';

/**
 * In this plugin, we do:
 * - Remove Babel loader if exists
 * - Add our own swc loader
 * - Remove JS minifier
 * - Add swc minifier plugin
 */
export const builderPluginSwc = (
  options: PluginSwcOptions = {},
): BuilderPlugin => ({
  name: PLUGIN_NAME,

  setup(api: BuilderPluginAPI) {
    if (api.context.bundlerType === 'rspack') {
      return;
    }

    api.modifyWebpackChain(async (chain, utils) => {
      const { CHAIN_ID, isProd } = utils;
      const builderConfig = api.getNormalizedConfig();
      const { rootPath } = api.context;

      const swcConfigs = await applyPluginConfig(
        options,
        utils,
        builderConfig,
        rootPath,
      );

      chain.module.rule(CHAIN_ID.RULE.JS).uses.delete(CHAIN_ID.USE.BABEL);
      chain.module.delete(CHAIN_ID.RULE.TS);

      const TJS_REGEX = mergeRegex(JS_REGEX, TS_REGEX);
      for (let i = 0; i < swcConfigs.length; i++) {
        const { test, include, exclude, swcConfig } = swcConfigs[i];

        const ruleId =
          i > 0 ? CHAIN_ID.RULE.JS + i.toString() : CHAIN_ID.RULE.JS;
        const rule = chain.module.rule(ruleId);

        // Insert swc loader and plugin
        rule
          .test(test || TJS_REGEX)
          .use(CHAIN_ID.USE.SWC)
          .loader(path.resolve(__dirname, './loader'))
          .options(removeUselessOptions(swcConfig) satisfies TransformConfig);

        if (include) {
          for (const extra of include) {
            rule.include.add(extra);
          }
        }

        if (exclude) {
          for (const extra of exclude) {
            rule.exclude.add(extra);
          }
        }
      }

      // first config is the main config
      const mainConfig = swcConfigs[0].swcConfig;

      if (chain.module.rules.get(CHAIN_ID.RULE.JS_DATA_URI)) {
        chain.module
          .rule(CHAIN_ID.RULE.JS_DATA_URI)
          .uses.delete(CHAIN_ID.USE.BABEL)
          .end()
          .use(CHAIN_ID.USE.SWC)
          .loader(path.resolve(__dirname, './loader'))
          .options(removeUselessOptions(mainConfig) satisfies TransformConfig);
      }

      if (isDebugMode()) {
        const { CheckPolyfillPlugin } = await import('./checkPolyfillPlugin');

        chain
          .plugin(CHAIN_ID.PLUGIN.SWC_POLYFILL_CHECKER)
          .use(new CheckPolyfillPlugin(mainConfig));
      }

      if (checkUseMinify(mainConfig, builderConfig, isProd)) {
        // Insert swc minify plugin
        // @ts-expect-error webpack-chain missing minimizers type
        const minimizersChain = chain.optimization.minimizers;

        if (mainConfig.jsMinify !== false) {
          minimizersChain.delete(CHAIN_ID.MINIMIZER.JS).end();
        }

        if (mainConfig.cssMinify !== false) {
          minimizersChain.delete(CHAIN_ID.MINIMIZER.CSS).end();
        }

        minimizersChain
          .end()
          .minimizer(CHAIN_ID.MINIMIZER.SWC)
          .use(SwcMinimizerPlugin, [
            {
              jsMinify: mainConfig.jsMinify ?? mainConfig.jsc?.minify,
              cssMinify: mainConfig.cssMinify,
              builderConfig,
            },
          ]);
      }
    });
  },
});

/// default swc configuration
export function getDefaultSwcConfig(): TransformConfig {
  const cwd = process.cwd();
  return {
    cwd,
    jsc: {
      externalHelpers: true,
      parser: {
        tsx: true,
        syntax: 'typescript',
        decorators: true,
      },
      transform: {
        react: {
          runtime: 'automatic',
        },
      },
      // Avoid the webpack magic comment to be removed
      // https://github.com/swc-project/swc/issues/6403
      preserveAllComments: true,
    },
    minify: false, // for loader, we don't need to minify, we do minification using plugin
    sourceMaps: true,
    env: {
      targets: '> 0.01%, not dead, not op_mini all',
    },
    exclude: [],
    inlineSourcesContent: true,
  };
}

/**
 * @deprecated Using builderPluginSwc instead.
 */
export const PluginSwc = builderPluginSwc;
