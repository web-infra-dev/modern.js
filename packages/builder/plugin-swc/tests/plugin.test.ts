import * as path from 'path';
import { expect, describe, it } from 'vitest';
import { createStubBuilder } from '@modern-js/builder-webpack-provider/stub';
import { builderPluginSwc } from '../src';
import { builderPluginBabel } from '@modern-js/builder-webpack-provider/plugins/babel';
import { createSnapshotSerializer } from '@scripts/vitest-config';

expect.addSnapshotSerializer(
  createSnapshotSerializer({
    replace: [{ mark: 'root', match: path.resolve(__dirname, '..') }],
  }),
);

describe('plugins/swc', () => {
  it('should set swc-loader', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginBabel(), builderPluginSwc()],
      builderConfig: {},
    });
    const config = await builder.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });

  it('should set swc minimizer in production', async () => {
    process.env.NODE_ENV = 'production';
    const builder = await createStubBuilder({
      plugins: [builderPluginBabel(), builderPluginSwc()],
      builderConfig: {},
    });
    const config = await builder.unwrapWebpackConfig();
    expect(config.optimization).toMatchSnapshot();

    process.env.NODE_ENV = 'test';
  });

  it('should disable swc minify', async () => {
    process.env.NODE_ENV = 'production';
    const builder = await createStubBuilder({
      plugins: [
        builderPluginBabel(),
        builderPluginSwc({
          jsMinify: false,
        }),
      ],
    });
    const config = await builder.unwrapWebpackConfig();
    expect(config.optimization).toBeFalsy();
    process.env.NODE_ENV = 'test';
  });

  it('should apply source.include and source.exclude correctly', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginBabel(), builderPluginSwc()],
      builderConfig: {
        source: {
          include: [/foo/],
          exclude: [/bar/],
        },
      },
    });
    const config = await builder.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });
});
