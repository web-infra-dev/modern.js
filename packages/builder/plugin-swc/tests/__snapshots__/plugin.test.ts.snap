// Vitest Snapshot v1

exports[`plugins/swc > should apply source.include and source.exclude correctly 1`] = `
{
  "module": {
    "rules": [
      {
        "exclude": [
          /bar/,
        ],
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /node_modules/,
              },
            ],
          },
          /foo/,
        ],
        "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$\\|\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/src/loader",
            "options": {
              "cwd": "<ROOT>",
              "env": {
                "coreJs": "3.26",
                "mode": "entry",
                "targets": [
                  "> 0.01%",
                  "not dead",
                  "not op_mini all",
                ],
              },
              "extensions": {
                "lockCorejsVersion": {
                  "corejs": "<WORKSPACE>/node_modules/<PNPM_INNER>/core-js",
                  "swcHelpers": "<WORKSPACE>/node_modules/<PNPM_INNER>/@swc/helpers",
                },
              },
              "jsc": {
                "transform": {
                  "react": {
                    "runtime": "classic",
                  },
                },
              },
            },
          },
        ],
      },
      {
        "mimetype": {
          "or": [
            "text/javascript",
            "application/javascript",
          ],
        },
        "use": [
          {
            "loader": "<ROOT>/src/loader",
            "options": {
              "cwd": "<ROOT>",
              "env": {
                "coreJs": "3.26",
                "mode": "entry",
                "targets": [
                  "> 0.01%",
                  "not dead",
                  "not op_mini all",
                ],
              },
              "extensions": {
                "lockCorejsVersion": {
                  "corejs": "<WORKSPACE>/node_modules/<PNPM_INNER>/core-js",
                  "swcHelpers": "<WORKSPACE>/node_modules/<PNPM_INNER>/@swc/helpers",
                },
              },
              "jsc": {
                "transform": {
                  "react": {
                    "runtime": "classic",
                  },
                },
              },
            },
          },
        ],
      },
    ],
  },
}
`;

exports[`plugins/swc > should set swc minimizer in production 1`] = `
{
  "minimizer": [
    SwcWebpackPlugin {
      "minifyOptions": {
        "compress": {},
        "mangle": true,
      },
    },
  ],
}
`;

exports[`plugins/swc > should set swc-loader 1`] = `
{
  "module": {
    "rules": [
      {
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /node_modules/,
              },
            ],
          },
        ],
        "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$\\|\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/src/loader",
            "options": {
              "cwd": "<ROOT>",
              "env": {
                "coreJs": "3.26",
                "mode": "entry",
                "targets": [
                  "> 0.01%",
                  "not dead",
                  "not op_mini all",
                ],
              },
              "extensions": {
                "lockCorejsVersion": {
                  "corejs": "<WORKSPACE>/node_modules/<PNPM_INNER>/core-js",
                  "swcHelpers": "<WORKSPACE>/node_modules/<PNPM_INNER>/@swc/helpers",
                },
              },
              "jsc": {
                "transform": {
                  "react": {
                    "runtime": "classic",
                  },
                },
              },
            },
          },
        ],
      },
      {
        "mimetype": {
          "or": [
            "text/javascript",
            "application/javascript",
          ],
        },
        "use": [
          {
            "loader": "<ROOT>/src/loader",
            "options": {
              "cwd": "<ROOT>",
              "env": {
                "coreJs": "3.26",
                "mode": "entry",
                "targets": [
                  "> 0.01%",
                  "not dead",
                  "not op_mini all",
                ],
              },
              "extensions": {
                "lockCorejsVersion": {
                  "corejs": "<WORKSPACE>/node_modules/<PNPM_INNER>/core-js",
                  "swcHelpers": "<WORKSPACE>/node_modules/<PNPM_INNER>/@swc/helpers",
                },
              },
              "jsc": {
                "transform": {
                  "react": {
                    "runtime": "classic",
                  },
                },
              },
            },
          },
        ],
      },
    ],
  },
}
`;
